Outline hexagonal architecture

mmjakub opened the pull request:
This is my attempt at moulding this CRUD app into hexagonal architecture. I skipped tests and adapters for now to streamline the review. Thanks to greater SoC this architecture could improve unit test performance, by using fake repositories (with hashmaps) instead of real SQLAlchemy sessions with inmemory sqlite.

mmjakub committed:
Outline hexagonal architecture

Jachu05 commented on app/core/models.py:
```diff
@@ -0,0 +1,25 @@
+    name: str
+    description: str
+    owner: str
+    collaborators: set[str]
```
it could be collaborators: set[User] or collaborators: List[User], instead of collaborators: set[str]. User would be more descriptive and it is allowed in domain. str is ambigous because it could be caseted id, name ect...

mmjakub replied:
```suggestion
    owner_usernames: str
    collaborator_usernames: set[str]
```
I think that OO approach is not needed here, as in project context we're only interested with usernames (for permission), not the whole User model. These field names should be more precise.

This solution avoids creating additional objects(Users) and I think is sufficient for our simple domain.

Jachu05 commented on app/core/ports/service.py:
```diff
@@ -0,0 +1,21 @@
+from typing import BinaryIO, Protocol
+
+
+class FileService(ABC):
```
lets make this a repository as this is also a data handler (simialr to sql but a file)

mmjakub replied:
```suggestion
class FileRepository(ABC):
```
Agreed, as discussed before.

mmjakub committed:
Apply proposed changes and improve file deletion

Renamed Project model fields to reflect that they reference usernames of Users
(owner and colaborators). I think this approach is better for now, because
we're not really dealing with User objects while handling Projects - just
veryfiyng usernames. So there are no OO benefits, just overhead costs.

Turned FileService into FileRepository as suggested.

Decided to call ports referring to outside services *Gateway instead of
*Service to avoid confusion with Application Layer services.

Introduced TaskQueueGateway, which for now is only used to perform file deletion
in background after serving the document deletion request. In future
this will also handle saving files, but requires api redesign (see #TODO).
For now let's see how this incremental change will impact testing, etc.

mmjakub commented on app/core/services.py:
```diff
@@ -20,22 +20,27 @@ async def _get_project_if_user_is_owner(
 
-async def _delete_document(document: Document, repository: DocumentRepository, file_service: FileService) -> None:
-    await asyncio.gather(repository.delete(document), file_service.delete(document.id))  # type: ignore
+async def _delete_document(
+    document: Document, document_repo: DocumentRepository, file_repo: FileRepository, task_queue: TaskQueueGateway
+) -> None:
+    await document_repo.delete(document)
+    if document.id is None:
+        raise RuntimeError("Document has no id")
+    task_queue.add_task(file_repo.delete, document.id)
```
As this simple domain has no data auditing/compliance requirements, Users are not really interested in what we do with the physical files after they delete the document. Using background task will greatly improve delete document request response times.

Jachu05 replied:
makes sense

mmjakub commented on app/core/services.py:
```diff
@@ -105,16 +111,18 @@ async def create_document(
     document = await document_repo.save(document)
-    # id: int | None causes problems here
-    await file_service.save(document.id, file)  # type: ignore
+    if document.id is None:
+        raise RuntimeError("Document has no id")
```
This is somewhat ugly(in my opinion) solution to typing errors caused by using`id: int | None`. This again touches on the problem of repository assigning ids to entities.

I thought of using some magic value (-1 seems to be good candidate) instead of None to indicate to the repository that the id should by generated by DBMS, but this smells like a hack to me. Would like to know your thoughts and best practices on this @Jachu05.

Jachu05 replied:
seems to be ok. here you do not need if document.id is None: as you are expecting await document_repo.save(document) to give you doc with id.

I will think about it more.

mmjakub commented on app/core/ports/gateway.py:
```diff
@@ -0,0 +1,13 @@
+from typing import Any, Callable, ParamSpec, Protocol
+
+
+class PasswordHashGateway(Protocol):
+    def hash(self, plain: str) -> str: ...
+    def verify(self, plain: str, hash: str) -> bool: ...
+
+
+P = ParamSpec("P")
+
+
+class TaskQueueGateway(Protocol):
+    def add_task(self, func: Callable[P, Any], *args: P.args, **kwargs: P.kwargs) -> int | None: ...
```
Renamed and moved outside services here.

mmjakub commented on app/core/models.py:
```diff
@@ -20,6 +20,6 @@ class Project:
-    owner: str
-    collaborators: set[str]
+    owner_username: str
+    collaborator_usernames: set[str]
```
> it could be collaborators: set[User] or collaborators: List[User], instead of collaborators: set[str]. User would be more descriptive and it is allowed in domain. str is ambigous because it could be caseted id, name ect...

I think that OO approach is not needed here, as in project context we're only interested with usernames (for permission), not the whole User model. I agree that these field names should be more precise.

This solution avoids creating additional objects(Users) and I think is sufficient for our simple domain.

Jachu05 replied:
in this case User is for readability, not because it is optimal.

I would say we either: make it User or create proper description/comment of attribute here with eventualy a docstrig

mmjakub committed:
Improve model readability by introducing Username type
